dhcpv4/options.go

```diff
@@ -85,70 +89,82 @@ func ParseOption(data []byte) (Option, error) {
 	case OptionVendorIdentifyingVendorClass:
 		opt, err = ParseOptVIVC(data)
 	default:
-		opt, err = ParseOptionGeneric(data)
+		opt, err = ParseOptionGeneric(code, data)
 	}
 	if err != nil {
 		return nil, err
 	}
 	return opt, nil
 }
 
-// OptionsFromBytes parses a sequence of bytes until the end and builds a list
-// of options from it. The sequence must contain the Magic Cookie. Returns an
-// error if any invalid option or length is found.
-func OptionsFromBytes(data []byte) ([]Option, error) {
-	if len(data) < len(MagicCookie) {
-		return nil, errors.New("invalid options: shorter than 4 bytes")
-	}
-	if !bytes.Equal(data[:len(MagicCookie)], MagicCookie) {
-		return nil, fmt.Errorf("invalid magic cookie: %v", data[:len(MagicCookie)])
-	}
-	opts, err := OptionsFromBytesWithoutMagicCookie(data[len(MagicCookie):])
-	if err != nil {
-		return nil, err
-	}
-	return opts, nil
-}
-
 // OptionsFromBytesWithoutMagicCookie parses a sequence of bytes until the end
 // and builds a list of options from it. The sequence should not contain the
 // DHCP magic cookie. Returns an error if any invalid option or length is found.
 func OptionsFromBytesWithoutMagicCookie(data []byte) ([]Option, error) {
-	return OptionsFromBytesWithParser(data, ParseOption)
+	return OptionsFromBytesWithParser(data, ParseOption, true)
 }
 
 // OptionParser is a function signature for option parsing
-type OptionParser func(data []byte) (Option, error)
+type OptionParser func(code OptionCode, data []byte) (Option, error)
 
 // OptionsFromBytesWithParser parses Options from byte sequences using the
 // parsing function that is passed in as a paremeter
-func OptionsFromBytesWithParser(data []byte, parser OptionParser) ([]Option, error) {
-	options := make([]Option, 0, 10)
-	idx := 0
-	for {
-		if idx == len(data) {
+func OptionsFromBytesWithParser(data []byte, parser OptionParser, checkEndOption bool) (Options, error) {
+	if len(data) == 0 {
+		return nil, nil
+	}
+	buf := uio.NewBigEndianBuffer(data)
+	options := make(map[OptionCode][]byte, 10)
+
+	// Due to RFC 3396 allowing an option to be specified multiple times,
+	// we have to collect all option data first, and then parse it.
+	var end bool
+	for buf.Len() >= 1 {
+		// 1 byte: option code
+		// 1 byte: option length n
+		// n bytes: data
+		code := OptionCode(buf.Read8())
+
+		if code == OptionPad {
+			continue
+		} else if code == OptionEnd {
+			end = true
 			break
 		}
-		// This should never happen.
-		if idx > len(data) {
-			return nil, errors.New("read past the end of options")
-		}
-		opt, err := parser(data[idx:])
-		idx++
-		if err != nil {
-			return nil, err
+		length := int(buf.Read8())
+
+		// N bytes: option data
+		data := buf.Consume(length)
+		if data == nil {
+			return nil, fmt.Errorf("error collecting options: %v", buf.Error())
 		}
-		options = append(options, opt)
-		if opt.Code() == OptionEnd {
-			break
+		data = data[:length:length]
```

Comment:
Is this different from `data = data[:length]`?

Reply:
Yes -- data is a slice into the overall array containing the packet. If you were to append more bytes here, you'd write over the packet if there are more bytes left in the packet. So you have to cap the capacity as well as the length.